src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp

```diff
@@ -0,0 +1,214 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "cds/aotClassLinker.hpp"
+#include "cds/aotLinkedClassBulkLoader.hpp"
+#include "cds/aotLinkedClassTable.hpp"
+#include "cds/cdsConfig.hpp"
+#include "classfile/classLoaderData.hpp"
+#include "classfile/systemDictionary.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "classfile/vmClasses.hpp"
+#include "gc/shared/gcVMOperations.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/instanceKlass.hpp"
+#include "oops/klass.inline.hpp"
+#include "runtime/handles.inline.hpp"
+#include "runtime/java.hpp"
+
+void AOTLinkedClassBulkLoader::serialize(SerializeClosure* soc, bool is_static_archive) {
+  AOTLinkedClassTable::get(is_static_archive)->serialize(soc);
+}
+
+void AOTLinkedClassBulkLoader::load_javabase_classes(JavaThread* current) {
+  assert(CDSConfig::is_using_aot_linked_classes(), "sanity");
+  load_classes_in_loader(current, AOTLinkedClassCategory::BOOT1, nullptr); // only java.base classes
+}
+
+void AOTLinkedClassBulkLoader::load_non_javabase_classes(JavaThread* current) {
+  assert(CDSConfig::is_using_aot_linked_classes(), "sanity");
+
+  // is_using_aot_linked_classes() requires is_using_full_module_graph(). As a result,
+  // the platform/system class loader should already have been initialized as part
+  // of the FMG support.
+  assert(CDSConfig::is_using_full_module_graph(), "must be");
+  assert(SystemDictionary::java_platform_loader() != nullptr, "must be");
+  assert(SystemDictionary::java_system_loader() != nullptr,   "must be");
+
+  load_classes_in_loader(current, AOTLinkedClassCategory::BOOT2, nullptr); // all boot classes outside of java.base
+  load_classes_in_loader(current, AOTLinkedClassCategory::PLATFORM, SystemDictionary::java_platform_loader());
+  load_classes_in_loader(current, AOTLinkedClassCategory::APP, SystemDictionary::java_system_loader());
+}
+
+void AOTLinkedClassBulkLoader::load_classes_in_loader(JavaThread* current, AOTLinkedClassCategory class_category, oop class_loader_oop) {
+  ExceptionMark em(current);
+  ResourceMark rm(current);
```

Comment:
The RM should go where it is actually needed for the logging.

Reply:
Fixed